### PR TITLE
use blacklist instead of nullary objects

### PR DESCRIPTION
**Pull request by eivindfjeldstad**
since setting `__proto__` doesn't work in v8 anymore.

### CODE REVIEW

**Comment by buschtoens** [comment]
I'm not a fan of these hackish attempts anyway. So I gladly welcome this PR.

I will give it a run through tomorrow (currently trying to fall asleep, haha) and merge it after.

**Comment by buschtoens** [comment]
At least Travis says all's well for now.

We should add tests for the blacklist.

**Comment by eivindfjeldstad** [comment]
[this](https://github.com/eivindfjeldstad/node-querystring/blob/60265e2f49a94bd0073d76d88c221a33ad111677/test/parse.js#L163) kinda covers it. but we could definitely add more tests

**Comment by buschtoens** [comment]
Oh, you _did_ add tests. :D

**Comment by buschtoens** [comment]
I see absolutely no problem with this PR. I'll test it tomorrow and add according notes to the readme.

Maybe @visionmedia likes to have a glance at this, since the proto stuff was added under his watch. :)

The only possible problem I see is that some crazy people maybe really would want to use the blacklisted names.

**Comment by eivindfjeldstad** [comment]
haha, there are always people doing crazy stuff. that said, you couldn't really use blacklisted names before either. they were overwritten when proto was restored.

**Comment by buschtoens** [comment]
Convinced. I'll still give TJ time til tomorrow to veto. Otherwise I'll assume he'll be fine with this and merge.

Thanks for you help. :)

**Comment by eivindfjeldstad** [comment]
cool! no problem:)

File: index.js [hunk 1]
```diff
@@ -110,6 +92,10 @@ function promote(parent, key) {
 
 function parse(parts, parent, key, val) {
   var part = parts.shift();
+  
+  // illegal
+  if (restricted[key]) return;
```
Comment (tj) [inline index.js]: might as well do a .hasOwnProperty or something here to account for falsey values

**Comment by tj** [comment]
other than the one comment LGTM

File: index.js (hunk 1 above)
Comment (eivindfjeldstad) [inline index.js]: `hasOwnProperty` will return false on prototypes, no? maybe

``` js
if (restricted[key] !== undefined) return;
```

Comment (tj) [inline index.js]: Object.getOwnPropertyDescriptor would be fine probably overkill though I suppose, no falsey values by default and people are exciting object.prototype then that sucks haha

Comment (eivindfjeldstad) [inline index.js]: I'll do `Object.getOwnPropertyDescriptor` if you're cool with that. You never know, people do some crazy stuff

**Comment by buschtoens** [comment]
Object.getOwnPropertyDescriptor is aupported in all major browsers and since IE8. Should be good enough. :)

**Comment by eivindfjeldstad** [comment]
yeah, I figured. want me to squash or is it good?

**Comment by buschtoens** [comment]
Would be quite nice, since the first commit alone doesn't do the trick. :)

**Comment by eivindfjeldstad** [comment]
done

**Comment by buschtoens** [comment]
All's well. Thanks again! :)

**Comment by eivindfjeldstad** [comment]
do you have time to push a new version to npm?

**Comment by buschtoens** [comment]
I would, [if I could](https://npmjs.org/package/qs). I already wrote to @visionmedia on [twitter](https://twitter.com/silvinci/status/407887472836612096). ;)

You can install qs [directly from the repo](https://npmjs.org/doc/install.html) for the time being.

**Comment by bajtos** [comment]
Well done! I came across this pull request while figuring out why `req.query` in my Express app has additional `__proto__` property when running on Node v0.11 (see joyent/node#6630).

@visionmedia When the new qs version is published, could you please update connect's package json to use this new version? It has a strict version specifier `0.6.5`.

**Pull request merged**